### ISSUE TITLE
Out-of-memory on TitanX (maxwell) w/ VGG16

### ISSUE THREAD

**Issue by drinkingkazu** [daijifeng001/MNC#30]
Hi,

I have just tried out MNC w/ VGG16 end-to-end training using pretty much out-of-the-box configuration (I believe).

I followed instructions to the point where I skipped running "Demo" in the README.
I went straight into "Training" after running 3 preparation steps to prepare data sample.
Then I ran the command suggested:
`cd $MNC_ROOT`
`./experiments/scripts/mnc_5stage.sh 0 VGG16`
The code runs to some point. In particular it construct the net fine.
But when it starts spitting out various task loss values @ iteration 0, it gets out-of-memory error.
I attach the full log file but here's the last lines that might be useful for a quick look.

`I1205 12:16:01.066789 22451 sgd_solver.cpp:106] Iteration 0, lr = 0.001`
`F1205 12:16:01.084414 22451 syncedmem.cpp:56] Check failed: error == cudaSuccess (2 vs. 0)  out of memory`
`*** Check failure stack trace: ***`
`./experiments/scripts/mnc_5stage.sh: line 35: 22451 Aborted                 (core dumped) ./tools/train_net.py --gpu ${GPU_ID} --solver models/${NET}/mnc_5stage/solver.prototxt --weights ${NET_INIT} --imdb ${DATASET_TRAIN} --iters ${ITERS} --cfg experiments/cfgs/${NET}/mnc_5stage.yml ${EXTRA_ARGS}`

**Comment by drinkingkazu** [comment]
I attach the log file (which I forgot in the previous post) and also a GPU monitoring plot that shows memory usage.

So the memory usage (cyan color) was near the max around 12:12, read off by my eyes.
The first timestamp from logger is around 12:10, and solving start around 12:11.
Given the accuracy of my eyes this matches w/ about when the gpu memory usage spike up to nearly full.
Then next timestamp kicks in around 12:16 in the log file, where it looks like reporting loss values.
I am assuming this is not from test net since I do not see test net config anywhere in yml nor solver prototxt.

It appears to me this is going out-of-mem when it tries to update the weights? or something.
The default configuration yml looks like it's already using mini-batch size = 1 so I assume this cannot be used to play around.
Since others seem to have no problem w/ VGG16 I'll try to debug myself and update here.
But anyone's input would be really appreciated.

![gpu2log_3600](https://cloud.githubusercontent.com/assets/4721630/20895290/70073af4-bade-11e6-9233-fb1ec9693b0f.png)

[log.txt](https://github.com/daijifeng001/MNC/files/631830/log.txt)

**Comment by Oh233** [comment]
Did you try to compile with cuDNN?

**Comment by drinkingkazu** [comment]
Yes, it is compiled w/ cuDNN (and opencv 3.1 are the modifications from default build config).
If this adds anything, the same configuration is used to build py-r-fcn on the same machine and that is training well w/ expected memory usage quoted in its README for ResNet-50, on the same machine.

**Comment by drinkingkazu** [comment]
Playing w/ iter_size in the solver.prototxt, I see computation time change linearly w/ iter_size as I expected (attached graph). One possibly interesting observation was that the crash was different when iter_size = 1. Instead of cuda out-of-mem error, I got Python MemoryError exception raised.

<img width="1084" alt="util_gpu2" src="https://cloud.githubusercontent.com/assets/4721630/20969964/0000e43e-bc51-11e6-9950-2a683ebbc7dd.png">

**Comment by drinkingkazu** [comment]
OK, I just happened to try on another gpu tower (there are 3) that does not have cudnn so I built w/o cudnn on that machine.
This tower is also maxwell TitanX (I briefly noted my failure on pascal TitanX that seems likely due to cudnn version #27).
Interestingly the training is proceeding fine on this machine... 
I will go back to the original machine I tested and see if this is the case (i.e. try building w/o cudnn).

**Comment by drinkingkazu** [comment]
Well this is embarrassing... but given the success on a separate machine I tried clean checkout and rebuilding on the original machine w/ cudnn, and now the training goes fine w/o any change in the codeset (to be clear using master branch  https://github.com/daijifeng001/MNC/commit/99410889179f2da71b9e3d4d6ceb8503371c5ab9).
I can see memory usage about 7.1 to 7.6 GB.

The source of the problem I saw earlier remains unknown but I am not interested in reproducing and digging in further for now.
Apology for the noise!
